Lesson20

ilya-frolov opened the pull request:
Task "Library" Mandatory part for review.

REMARK: I initially provided random 'id' creation, however it wasn't convenient to perform verification of all methods, so this part of code is commented.

ilya-frolov committed:
lesson20: Task "Library" is completed.

ilya-frolov committed:
Delete Lesson 1 - HelloWorld directory

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Book.java:
```diff
@@ -0,0 +1,73 @@
+
+    public Book(String title, int id) {
+//        Random random = new Random();
+//        id = random.nextInt(999) + 1;
```
remove comments from MRs

ilya-frolov replied:
Done.

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Book.java:
```diff
@@ -0,0 +1,73 @@
+    private String genre;
+    private Date date;
+
+    public Book(String title, int id) {
```
let's generate ids automatically (every new book should have id greater than previous on 1)

ilya-frolov replied:
Done.

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Book.java:
```diff
@@ -0,0 +1,73 @@
+//        id = random.nextInt(999) + 1;
+        this.id = id;
+        this.title = title;
+        this.genre = new Genre().getGenre(id);
```
nonono. Genre should be passes into the constructor - don't generate genre almost randomly

ilya-frolov replied:
Ok.

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Book.java:
```diff
@@ -0,0 +1,73 @@
+    // For case of List Library
+    @Override
+    public int compareTo(Book book) {
+        if (getTitle() == null || book.getTitle() == null) {
```
so, if one book has title and other one not - they should be equal? I don't think it's correct

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Genre.java:
```diff
@@ -0,0 +1,20 @@
+package com.ilyafrolov.lesson20;
+
+public class Genre {
```
I prefere enum for Genres (it's expected that list of genres will not be changed so often)

ilya-frolov replied:
Done.

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Library.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    public void addBook(Book book) {
+        boolean isUnique = true;
```
books can not be unique (almost all libraries has a lot of similar books). I would suggest you to store books inside hash map by their title for example (you could have books with same title, but with different ids, or even genres). But it leads to another problem - if you want to get/remove book by ID you need to flattern all your map and look for the book with required ID. So, you should think how it would be better to act in this case

ilya-frolov replied:
Why I cannot use ID as a Key?

ilya-frolov replied:
Done.

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Library.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    public void sortByTitle() {
+        Collections.sort(books);
```
It's not obvious that sorting will be made by title. It's better to create here comparator (or even better to create comparator at the class level (to be created just once) and use it here - similar with sorting by date but with comparator creation on class level

ilya-frolov replied:
Done.

ilya-frolov committed:
lesson20: Task "Library":
- changed data structure from LinkedList to HashMap;
- changed receiving id operation;
- made Genre enum;
- made Comparator classes in class Book and methods for comparing in class Library.

ilya-frolov committed:
Merge remote-tracking branch 'origin/lesson20' into lesson20

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Book.java:
```diff
@@ -62,12 +61,24 @@ public int hashCode() {
-    public int compareTo(Book book) {
-        if (getTitle() == null || book.getTitle() == null) {
-            return 0;
+    public static class sortByTitle implements Comparator<String> {
```
Ненене, компаратор не может архитектурно находиться в книге - он должен быть в библиотеке. + имена классов с большой буквы

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Book.java:
```diff
@@ -62,12 +61,24 @@ public int hashCode() {
+        }
+    }
+
+    public static class sortByTitleReverse implements Comparator<String> {
```
Ты уже отсортировал по titlу, так что не надо второй раз - вызываешь предыдущий и просто выводишь в обратном порядке (или делаешь уже на месте  reverse

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Book.java:
```diff
@@ -62,12 +61,24 @@ public int hashCode() {
+        }
+    }
+
+    public static class sortByDate implements Comparator<Book> {
```
ну как и выше

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Book.java:
```diff
@@ -1,29 +1,29 @@
         return id;
     }
 
+    public void setId(int id) {
```
нет. Убирай метод. Мы не должны иметь возможость менять id в книге.

ilya-frolov replied:
Точняк. Done.

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Book.java:
```diff
@@ -1,29 +1,29 @@
-public class Book implements Comparable<Book> {
+public class Book {
 
     private int id;
```
private final

ilya-frolov replied:
Done.

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Library.java:
```diff
@@ -4,65 +4,87 @@
 
-    List<Book> books;
+    private Map<String, ArrayList<Book>> books;
+    private ArrayList<Book> list;
```
убирай

ilya-frolov replied:
А он мне нужен, я в него загоняю несколько одинаковых книг, а он в свою очередь лежит в общей коллекции books.values(). Единственное, что я его перенесу внуть метода addBook();

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Library.java:
```diff
@@ -4,65 +4,87 @@
-    List<Book> books;
+    private Map<String, ArrayList<Book>> books;
+    private ArrayList<Book> list;
+    private ArrayList<String> titleList;
```
убирай

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Library.java:
```diff
@@ -4,65 +4,87 @@
-        boolean isUnique = true;
-        if (books.size() == 0) {
-            books.add(book);
+        book.setId(books.size() + 7);
```
нет, плохо. убирай и сделай генерацию id в конструкторе Book

ilya-frolov replied:
Done.

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Library.java:
```diff
@@ -4,65 +4,87 @@
-            books.add(book);
+        book.setId(books.size() + 7);
+        if (books.containsKey(book.getTitle())) {
+            books.get(book.getTitle()).add(book);
```
попробуй метод compute(...) в HashMap

ilya-frolov replied:
Метод compute(...) заменяет существующее значение, мне же здесь наоборот нужно добавлять к существующей в Листе книге еще одну такую же, но с другим id. 
Плюс этот метод использует Лямбда выражения, которые мы не учили еще (хоть здесь оно и банальное).

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Library.java:
```diff
@@ -4,65 +4,87 @@
-        for (int i = 0; i < books.size(); i++) {
-            if (books.get(i).getId() == id) {
-                books.remove(books.get(i));
+        titleList = new ArrayList(books.keySet());
```
очень сложно. можно же просто бежать по books.values() - это и будет коллекция List'ов книг

ilya-frolov replied:
Чтобы бежать по Коллекции books.values() я нашел два способа.
1. Iterator, но не смог победить отсутствие обнуления счетчика после перехода на новый iterator.next().
2. Сделал с помощью for-each, по сути тоже самое, что у меня было, только запись более красивая. Только теперь вместо одного Листа для titles, каждый раз создается ArrayList для каждого элемента Коллекции, если я правильно понимаю.

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Library.java:
```diff
@@ -4,65 +4,87 @@
-        for (int i = 0; i < books.size(); i++) {
-            if (books.get(i).getId() == id) {
-                books.get(i).setTitle(newTitle);
+        titleList = new ArrayList(books.keySet());
```
аналогично предыдущему

ilya-frolov replied:
Сделал с помощью for-each

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Library.java:
```diff
@@ -4,65 +4,87 @@
 
+    public void sortByTitle() {
+        titleList = new ArrayList(books.keySet());
+        Collections.sort(titleList, new Book.sortByTitle());
```
а был бы компаратор тут - не пришлось бы на каждый вызов создавать новый объект компаратора

ilya-frolov replied:
Не понимаю, мы же в любом случае должны создать объект компаратор, который будет сравнивать 2 объекта. Или я что-то не так понимаю. Я то перенес класс Компаратора в Library, но его объект я же должен создавать.

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Library.java:
```diff
@@ -4,65 +4,87 @@
+
+    public void sortByTitleReverse() {
+        titleList = new ArrayList(books.keySet());
+        Collections.sort(titleList, new Book.sortByTitleReverse());
```
см выше

RomanKashenok commented on HomeWorkTask20_Library/src/com/ilyafrolov/lesson20/Library.java:
```diff
@@ -4,65 +4,87 @@
             }
-        });
+        }
+        Collections.sort(list, new Book.sortByDate());
```
см выше

ilya-frolov replied:
Ты не делал замечание по самому методу здесь, но он все равно громоздкий. Этот метод никак я не смог упростить. Чуть переделал логику, но по сути то же самое осталось. Пробовал Map.Entry для использовать, но так и не победил получение доступа к каждой книге, находящейся в Листе, находищемся в общей коллекции books.values().

ilya-frolov committed:
lesson20: Task "Library":
- changed data structure from HashMap to TreeMap;
- changed receiving id operation;
- made Comparator class in Library class;
- moved some parameters from general variables to local variables, deleted some Lists;
- changed method for comparing by date.

ilya-frolov committed:
Delete BookOldList.java
